### PR TITLE
MDS-689: Associate Parties with OrgBook Entities

### PR DESCRIPTION
**Pull request by exzizt**
# Main
* Allows users to view and associate parties to an entity on OrgBook
  * This is done via the **Update Party** modal:
image
  * Use the search to locate an entity on OrgBook, select it, then confirm any additional details using the **View on OrgBook** button. Once you've verified, select the **Associate** button to associate this party with the OrgBook entity:
image
image
  * You can view whether or not a party has an associated OrgBook entity by looking at its contact information on the contact page or by looking for the green checkmark beside its name in the contacts table (for now):
image
image

# Other
* Fixes a bug where refreshing/loading the Contacts page would crash, due to a "party selector" error

# Notes
Request/response samples to the two OrgBook API calls:

1. `https://orgbook.gov.bc.ca/api/v2/search/autocomplete?q=zinex`
```
{
   "total":1,
   "first_index":1,
   "last_index":1,
   "results":[
      {
         "id":777855,
         "names":[
            {
               "id":712594,
               "text":"ZINEX MINING CORP.",
               "language":null,
               "credential_id":777855,
               "type":"entity_name"
            }
         ],
         "inactive":false
      }
   ]
}
```

2. `https://orgbook.gov.bc.ca/api/v2/credential/777855/formatted`
```
{
   "id":777855,
   "create_timestamp":"2019-06-25T23:01:10.740961-07:00",
   "effective_date":"2006-11-10T14:18:19-08:00",
   "inactive":false,
   "latest":true,
   "revoked":false,
   "revoked_date":null,
   "wallet_id":"92101ca6-1c55-4078-8146-855a1afb613c",
   "credential_type":{
      "id":1,
      "issuer":{
         "id":1,
         "has_logo":true,
         "create_timestamp":"2019-06-25T14:52:20.353243-07:00",
         "update_timestamp":"2020-05-06T13:01:41.230287-07:00",
         "did":"HR6vs6GEZ8rHaVgjg2WodM",
         "name":"BC Corporate Registry",
         "abbreviation":"BCReg",
         "email":"bcregistries@gov.bc.ca",
         "url":"https://www2.gov.bc.ca/gov/content/governments/organizational-structure/ministries-organizations/ministries/citizens-services/bc-registries-online-services",
         "endpoint":null
      },
      "has_logo":true,
      "create_timestamp":"2019-06-25T14:52:20.428303-07:00",
      "update_timestamp":"2020-05-06T14:30:56.182057-07:00",
      "description":"Registration",
      "credential_def_id":"HR6vs6GEZ8rHaVgjg2WodM:3:CL:41051:tag",
      "last_issue_date":"2020-05-06T14:30:56.181965-07:00",
      "url":"https://bcreg-x-proxy-devex-von-bc-registries-agent-prod.pathfinder.gov.bc.ca/bcreg/incorporation",
      "schema":{
         "id":1,
         "create_timestamp":"2019-06-25T14:52:20.397843-07:00",
         "update_timestamp":"2020-05-06T13:01:41.243388-07:00",
         "name":"registration.registries.ca",
         "version":"1.0.42",
         "origin_did":"HR6vs6GEZ8rHaVgjg2WodM"
      }
   },
   "addresses":[

   ],
   "attributes":[
      {
         "id":5381558,
         "type":"registration_date",
         "format":"datetime",
         "value":"2006-11-10T22:18:19+00:00",
         "credential_id":777855
      },
      {
         "id":5381559,
         "type":"entity_name_effective",
         "format":"datetime",
         "value":"2006-11-10T22:18:19+00:00",
         "credential_id":777855
      },
      {
         "id":5381560,
         "type":"entity_status",
         "format":"category",
         "value":"ACT",
         "credential_id":777855
      },
      {
         "id":5381561,
         "type":"entity_status_effective",
         "format":"datetime",
         "value":"2006-11-10T22:18:19+00:00",
         "credential_id":777855
      },
      {
         "id":5381562,
         "type":"entity_type",
         "format":"category",
         "value":"BC",
         "credential_id":777855
      },
      {
         "id":5381563,
         "type":"home_jurisdiction",
         "format":"jurisdiction",
         "value":"BC",
         "credential_id":777855
      },
      {
         "id":5381564,
         "type":"reason_description",
         "format":"category",
         "value":"Filing:ICORP",
         "credential_id":777855
      }
   ],
   "names":[
      {
         "id":712594,
         "text":"ZINEX MINING CORP.",
         "language":null,
         "credential_id":777855,
         "type":"entity_name"
      }
   ],
   "local_name":{
      "id":712594,
      "text":"ZINEX MINING CORP.",
      "language":null,
      "credential_id":777855,
      "type":"entity_name"
   },
   "remote_name":null,
   "topic":{
      "id":683004,
      "create_timestamp":"2019-06-25T23:01:10.721202-07:00",
      "update_timestamp":"2019-06-25T23:01:10.721249-07:00",
      "source_id":"BC0774378",
      "type":"registration",
      "names":[
         {
            "id":712594,
            "text":"ZINEX MINING CORP.",
            "language":null,
            "credential_id":777855,
            "type":"entity_name"
         }
      ],
      "local_name":{
         "id":712594,
         "text":"ZINEX MINING CORP.",
         "language":null,
         "credential_id":777855,
         "type":"entity_name"
      },
      "remote_name":null,
      "addresses":[

      ],
      "attributes":[
         {
            "id":5381558,
            "type":"registration_date",
            "format":"datetime",
            "value":"2006-11-10T22:18:19+00:00",
            "credential_id":777855,
            "credential_type_id":1
         },
         {
            "id":5381559,
            "type":"entity_name_effective",
            "format":"datetime",
            "value":"2006-11-10T22:18:19+00:00",
            "credential_id":777855,
            "credential_type_id":1
         },
         {
            "id":5381560,
            "type":"entity_status",
            "format":"category",
            "value":"ACT",
            "credential_id":777855,
            "credential_type_id":1
         },
         {
            "id":5381561,
            "type":"entity_status_effective",
            "format":"datetime",
            "value":"2006-11-10T22:18:19+00:00",
            "credential_id":777855,
            "credential_type_id":1
         },
         {
            "id":5381562,
            "type":"entity_type",
            "format":"category",
            "value":"BC",
            "credential_id":777855,
            "credential_type_id":1
         },
         {
            "id":5381563,
            "type":"home_jurisdiction",
            "format":"jurisdiction",
            "value":"BC",
            "credential_id":777855,
            "credential_type_id":1
         },
         {
            "id":5381564,
            "type":"reason_description",
            "format":"category",
            "value":"Filing:ICORP",
            "credential_id":777855,
            "credential_type_id":1
         }
      ]
   },
   "related_topics":[

   ],
   "credential_set":{
      "id":708701,
      "create_timestamp":"2019-06-25T23:01:10.780834-07:00",
      "update_timestamp":"2019-06-25T23:01:10.780864-07:00",
      "latest_credential_id":777855,
      "topic_id":683004,
      "first_effective_date":"2006-11-10T14:18:19-08:00",
      "last_effective_date":null,
      "credentials":[
         {
            "id":777855,
            "create_timestamp":"2019-06-25T23:01:10.740961-07:00",
            "effective_date":"2006-11-10T14:18:19-08:00",
            "inactive":false,
            "latest":true,
            "revoked":false,
            "revoked_date":null,
            "wallet_id":"92101ca6-1c55-4078-8146-855a1afb613c",
            "credential_type":{
               "id":1,
               "issuer":{
                  "id":1,
                  "has_logo":true,
                  "create_timestamp":"2019-06-25T14:52:20.353243-07:00",
                  "update_timestamp":"2020-05-06T13:01:41.230287-07:00",
                  "did":"HR6vs6GEZ8rHaVgjg2WodM",
                  "name":"BC Corporate Registry",
                  "abbreviation":"BCReg",
                  "email":"bcregistries@gov.bc.ca",
                  "url":"https://www2.gov.bc.ca/gov/content/governments/organizational-structure/ministries-organizations/ministries/citizens-services/bc-registries-online-services",
                  "endpoint":null
               },
               "has_logo":true,
               "create_timestamp":"2019-06-25T14:52:20.428303-07:00",
               "update_timestamp":"2020-05-06T14:30:56.182057-07:00",
               "description":"Registration",
               "credential_def_id":"HR6vs6GEZ8rHaVgjg2WodM:3:CL:41051:tag",
               "last_issue_date":"2020-05-06T14:30:56.181965-07:00",
               "url":"https://bcreg-x-proxy-devex-von-bc-registries-agent-prod.pathfinder.gov.bc.ca/bcreg/incorporation",
               "schema":{
                  "id":1,
                  "create_timestamp":"2019-06-25T14:52:20.397843-07:00",
                  "update_timestamp":"2020-05-06T13:01:41.243388-07:00",
                  "name":"registration.registries.ca",
                  "version":"1.0.42",
                  "origin_did":"HR6vs6GEZ8rHaVgjg2WodM"
               }
            },
            "addresses":[

            ],
            "attributes":[
               {
                  "id":5381558,
                  "type":"registration_date",
                  "format":"datetime",
                  "value":"2006-11-10T22:18:19+00:00",
                  "credential_id":777855
               },
               {
                  "id":5381559,
                  "type":"entity_name_effective",
                  "format":"datetime",
                  "value":"2006-11-10T22:18:19+00:00",
                  "credential_id":777855
               },
               {
                  "id":5381560,
                  "type":"entity_status",
                  "format":"category",
                  "value":"ACT",
                  "credential_id":777855
               },
               {
                  "id":5381561,
                  "type":"entity_status_effective",
                  "format":"datetime",
                  "value":"2006-11-10T22:18:19+00:00",
                  "credential_id":777855
               },
               {
                  "id":5381562,
                  "type":"entity_type",
                  "format":"category",
                  "value":"BC",
                  "credential_id":777855
               },
               {
                  "id":5381563,
                  "type":"home_jurisdiction",
                  "format":"jurisdiction",
                  "value":"BC",
                  "credential_id":777855
               },
               {
                  "id":5381564,
                  "type":"reason_description",
                  "format":"category",
                  "value":"Filing:ICORP",
                  "credential_id":777855
               }
            ],
            "names":[
               {
                  "id":712594,
                  "text":"ZINEX MINING CORP.",
                  "language":null,
                  "credential_id":777855,
                  "type":"entity_name"
               }
            ],
            "local_name":{
               "id":712594,
               "text":"ZINEX MINING CORP.",
               "language":null,
               "credential_id":777855,
               "type":"entity_name"
            },
            "remote_name":null,
            "topic":{
               "id":683004,
               "create_timestamp":"2019-06-25T23:01:10.721202-07:00",
               "update_timestamp":"2019-06-25T23:01:10.721249-07:00",
               "source_id":"BC0774378",
               "type":"registration",
               "names":[
                  {
                     "id":712594,
                     "text":"ZINEX MINING CORP.",
                     "language":null,
                     "credential_id":777855,
                     "type":"entity_name"
                  }
               ],
               "local_name":{
                  "id":712594,
                  "text":"ZINEX MINING CORP.",
                  "language":null,
                  "credential_id":777855,
                  "type":"entity_name"
               },
               "remote_name":null,
               "addresses":[

               ],
               "attributes":[
                  {
                     "id":5381558,
                     "type":"registration_date",
                     "format":"datetime",
                     "value":"2006-11-10T22:18:19+00:00",
                     "credential_id":777855,
                     "credential_type_id":1
                  },
                  {
                     "id":5381559,
                     "type":"entity_name_effective",
                     "format":"datetime",
                     "value":"2006-11-10T22:18:19+00:00",
                     "credential_id":777855,
                     "credential_type_id":1
                  },
                  {
                     "id":5381560,
                     "type":"entity_status",
                     "format":"category",
                     "value":"ACT",
                     "credential_id":777855,
                     "credential_type_id":1
                  },
                  {
                     "id":5381561,
                     "type":"entity_status_effective",
                     "format":"datetime",
                     "value":"2006-11-10T22:18:19+00:00",
                     "credential_id":777855,
                     "credential_type_id":1
                  },
                  {
                     "id":5381562,
                     "type":"entity_type",
                     "format":"category",
                     "value":"BC",
                     "credential_id":777855,
                     "credential_type_id":1
                  },
                  {
                     "id":5381563,
                     "type":"home_jurisdiction",
                     "format":"jurisdiction",
                     "value":"BC",
                     "credential_id":777855,
                     "credential_type_id":1
                  },
                  {
                     "id":5381564,
                     "type":"reason_description",
                     "format":"category",
                     "value":"Filing:ICORP",
                     "credential_id":777855,
                     "credential_type_id":1
                  }
               ]
            },
            "related_topics":[

            ]
         }
      ]
   }
}
```

### CODE REVIEW

File: .vscode/.ropeproject/config.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
+# The default ``config.py``
```
Comment (exzizt) [inline .vscode/.ropeproject/config.py:1]: This should probably be added to .gitignore... unless we want this.

File: services/core-api/app/api/services/orgbook_service.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import requests
+
+from app.config import Config
+
+
+class OrgBookService():
```
Comment (nayr974) [inline services/core-api/app/api/services/orgbook_service.py:6]: Nice; I like this approach

File: services/core-web/common/actionCreators/orgbookActionCreator.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    .get(ENVIRONMENT.apiUrl + API.ORGBOOK_SEARCH(search), createRequestHeader())
+    .then((response) => {
+      dispatch(success(reducerTypes.ORGBOOK_SEARCH));
+      dispatch(orgbookActions.storeSearchOrgBookResults(response.data));
+    })
+    .catch(() => dispatch(error(reducerTypes.ORGBOOK_SEARCH)))
+    .finally(() => dispatch(hideLoading()));
+};
+
+export const fetchOrgBookCredential = (credentialId) => (dispatch) => {
```
Comment (nayr974) [inline services/core-web/common/actionCreators/orgbookActionCreator.js:23]: I thought we were going through our API as a proxy for calls to orgbook. Why does the front-end need a credential?

Comment (exzizt) [reply services/core-web/common/actionCreators/orgbookActionCreator.js:23]: The search field uses a search endpoint that doesn't return anything useful except the name of the entity and its latest credential ID. Using the latest credential ID, we can make an API call (fetchOrgBookCredential) to get it's source_id, which is needed to create the URL for the **View on OrgBook** button, as well as to display any other additional info we may want to show the user in the UI before they confirm and click the **Associate** button.

File: .vscode/.ropeproject/config.py (hunk 1 above)
Comment (exzizt) [reply .vscode/.ropeproject/config.py]: Deleted it

File: services/core-web/common/actionCreators/orgbookActionCreator.js (hunk 3 above)
Comment (pdittaro) [reply services/core-web/common/actionCreators/orgbookActionCreator.js:23]: Credential is VON name for the record, not a credential for the actual web service.

File: services/core-api/app/config.py [hunk 4]
```diff
@@ -103,15 +103,18 @@ def JWT_ROLE_CALLBACK(jwt_dict):
...
     VFCBC_CLIENT_ID = os.environ.get('VFCBC_CLIENT_ID', None)
 
-    # NRIS Remote API
+    # NRIS
     NRIS_REMOTE_CLIENT_SECRET = os.environ.get('NRIS_REMOTE_CLIENT_SECRET', None)
     NRIS_REMOTE_CLIENT_ID = os.environ.get('NRIS_REMOTE_CLIENT_ID', None)
     NRIS_REMOTE_TOKEN_URL = os.environ.get('NRIS_REMOTE_TOKEN_URL', None)
 
+    # OrgBook
+    ORGBOOK_API_URL = 'https://orgbook.gov.bc.ca/api/v2/'
```
Comment (pdittaro) [inline services/core-api/app/config.py:116]: Can we have this as an environment variable as well? Good practice, plus once we become an issuer we might be running a local orgbook instance for development.

Comment (exzizt) [reply services/core-api/app/config.py:116]: Sure, although, when we are running our own local instance I think we will still need this URL in order to retrieve data stored in VON/OrgBook.

**Pull request merged**